frame/evm/src/runner/meter.rs

```diff
@@ -0,0 +1,155 @@
+use evm::{gasometer::GasCost, Opcode};
+use fp_evm::ACCOUNT_STORAGE_PROOF_SIZE;
+use sp_core::H256;
+
+/// A meter for tracking the storage growth.
+#[derive(Clone, Copy)]
+pub struct StorageMeter {
+	usage: u64,
+	limit: u64,
+}
+
+/// An error that is returned when the storage limit has been exceeded.
+#[derive(Debug, PartialEq)]
+pub enum MeterError {
+	LimitExceeded,
+}
+
+impl StorageMeter {
+	/// Creates a new storage meter with the given limit.
+	pub fn new(limit: u64) -> Self {
+		Self { usage: 0, limit }
+	}
+
+	/// Records the given amount of storage usage. The amount is added to the current usage.
+	/// If the limit is reached, an error is returned.
+	pub fn record(&mut self, amount: u64) -> Result<(), MeterError> {
+		self.usage = self.usage.checked_add(amount).ok_or_else(|| {
+			self.usage = self.limit;
+			MeterError::LimitExceeded
+		})?;
+
+		if self.usage > self.limit {
+			return Err(MeterError::LimitExceeded);
+		}
+		Ok(())
+	}
+
+	/// Records the storage growth for the given Opcode.
+	pub fn record_dynamic_opcode_cost(
+		&mut self,
+		_opcode: Opcode,
+		gas_cost: GasCost,
+	) -> Result<(), MeterError> {
+		match gas_cost {
+			GasCost::SStore { original, new, .. }
+				if original == H256::default() && !new.is_zero() =>
```

Comment:
Would it be equivalent to `if new != original`?

Reply:
No, they're not the same. Consider this specific case:

- **Case of Difference**: When `original` is `5` and `new` is `0`, the first condition evaluates to `FALSE`, whereas the second condition evaluates to `TRUE`.

The first condition is designed to catch a transition from an uninitialized or zero state to a non-zero state, something that the second condition doesn't specifically account for.